Lets encrypt

mbabauer opened the pull request:
Updates to allow Let's Encrypt Docker Image to be used for auto-generated certificates.

arunoda commented:
okay. I like this. 
Shall we add this to the docs. 
And possibly, into the conman like help as well.

arunoda commented:
We've merged a similar PR into mupx. 
Thanks for your PR.
But it's the same functionality.
See: https://github.com/arunoda/meteor-up/pull/417

Commit:
Initial attempt at Let's Encrypt

Commit:
Fixes for nginx docker volumes

Commit:
Latest updates to support lets encrypt

Commit:
Changed the separated nginx and docker-gen containers to use the combined nginx-proxy.

Commit:
Additional cleanup of commented out code.

brylie commented:
@mbabauer are the conflicts here difficult to resolve?

s-devaney commented:
+1

mbabauer committed:
Merge pull request #1 from kadirahq/master

Let's Encrypt Docker Proxy

Commit:
Merge remote-tracking branch 'origin' into lets-encrypt

brylie commented:
@madushan1000, may this please be reviewed and/or merged soon?

brylie commented on src/modules/meteor/assets/templates/start.sh:
```diff
@@ -29,21 +31,70 @@ docker run \
   <% for(var volume in volumes) { %>-v <%= volume %>:<%= volumes[volume] %> <% } %>\
   <% for(var args in docker.args) { %> <%= docker.args[args] %> <% } %>\
+  <% if(typeof sslConfig.autogenerate === "object")  { %> \
+    -e "VIRTUAL_HOST=proclaim.32past8.com" \
```
Should `proclaim.32past8.com` be hard coded here?

mbabauer replied:
Sorry, that's my host I was using to test with.  I'll update the PR to remove that.

brylie reviewed (commented):
What is the `proclaim.32past8.com` string used for here?

brylie commented:
@mbabauer does this PR auto-renew the LetsEncrypt certificate(s)?

came commented:
+1

mbabauer commented:
@brylie The docker image I am using is [here](https://github.com/JrCs/docker-letsencrypt-nginx-proxy-companion).  According to their documentation, it should auto-renew the certs.  It hasn't been 90 days since my instance was started, though, so I can't tell you with 100% certainty.

Commit:
Fixing hostname issue

kulttuuri commented:
+1

brylie commented:
@mbabauer, conflicts have been resolved and the hardcoded URL was removed. Any timeline for further review of this PR?

goooseman commented:
there is a little bug.

if there is no ssl section in deploy.js it is throwing an error:

```
[xxx.xxx.xxx.xxx] - Pushing the Startup Script
TypeError ejs:34
    32|   <% for(var volume in volumes) { %>-v <%= volume %>:<%= volumes[volume] %> <% } %>\
    33|   <% for(var args in docker.args) { %> <%= docker.args[args] %> <% } %>\
 >> 34|   <% if(typeof sslConfig.autogenerate === "object")  { %> \
    35|     -e "VIRTUAL_HOST=$HOSTNAME" \
    36|     -e "LETSENCRYPT_HOST=<%= sslConfig.autogenerate.domains %>" \
    37|     -e "LETSENCRYPT_EMAIL=<%= sslConfig.autogenerate.email %>" \

Cannot read property 'autogenerate' of undefined
```

brylie commented:
@goooseman good catch. @mbabauer would this be hard to fix?

goooseman commented:
@brylie 
in start.sh change line no 34 to:
```
  <% if(sslConfig && typeof sslConfig.autogenerate === "object")  { %> \
```

testing right now and will make a PR

Commit:
[FIX] Check for sslConfig

Previously, if there was no ssl section in deploy.js it was throwing an error:
```
[xxx.xxx.xxx.xxx] - Pushing the Startup Script
TypeError ejs:34
    32|   <% for(var volume in volumes) { %>-v <%= volume %>:<%= volumes[volume] %> <% } %>\
    33|   <% for(var args in docker.args) { %> <%= docker.args[args] %> <% } %>\
 >> 34|   <% if(typeof sslConfig.autogenerate === "object")  { %> \
    35|     -e "VIRTUAL_HOST=$HOSTNAME" \
    36|     -e "LETSENCRYPT_HOST=<%= sslConfig.autogenerate.domains %>" \
    37|     -e "LETSENCRYPT_EMAIL=<%= sslConfig.autogenerate.email %>" \

Cannot read property 'autogenerate' of undefined
```

I have added a check for sslConfig, so now it works even when there is no ssl section in deploy.js

goooseman commented:
https://github.com/mbabauer/meteor-up/pull/2

Commit:
Merge pull request #1 from goooseman/patch-1

[FIX] Check for sslConfig

Commit:
[FIX] Wasn't working without ssl

When we are using autogenerate, we are exposing 80 port.
When we are not using autogenerate, we should publish 80 port.

Commit:
Merge pull request #2 from goooseman/patch-2

[FIX] Wasn't working without ssl

goooseman commented:
Found one more bug, when library was used without ssl. Fixed. 
https://github.com/mbabauer/meteor-up/pull/3

goooseman commented:
But with ssl autogenerate object it doesn't work for me.

It just show always 503

goooseman commented:
f## it. i'm leaving.

what i have done at the end. i installed acmetool manually on the server. he checks for the new cert. and i have made a symlink to the /opt/projectName/config folder:

```
$APP_NAME=test
$HOST=test.discours.io
sudo add-apt-repository ppa:hlandau/rhea
sudo apt-get update
sudo apt-get install acmetool
sudo acmetool quickstart
sudo acmetool want $HOST
mkdir -p /opt/$APP_NAME/config/
rm /opt/$APP_NAME/config/bundle.crt /opt/$APP_NAME/config/private.key
ln -s /var/lib/acme/live/$HOST/fullchain /opt/$APP_NAME/config/bundle.crt
ln -s /var/lib/acme/live/$HOST/privkey /opt/$APP_NAME/config/private.key

```

mbabauer committed:
Merge pull request #3 from goooseman/lets-encrypt

[FIX] working without autogenerate

brylie commented:
@madushan1000 please review this PR, and advise of any needed changes.

RemyLespagnol commented:
+1 👍

lucbocahut commented:
I got this to work by changing line 39 to 
-e "VIRTUAL_HOST=<%= sslConfig.autogenerate.domains %>" \
instead of 
-e "VIRTUAL_HOST=$HOSTNAME" \

lucbocahut commented:
If its unclear how to get this to work:
`git clone https://github.com/mbabauer/meteor-up.git#lets-encrypt` clones the repo to your current local directory
edit meteor-up/src/modules/meteor/templates/start.sh as indicated above by changing VIRTUAL_HOST=<%= sslConfig.autogenerate.domains %> (will only work if you set autogenerate in mup.js)
`cd meteor-up` cd inside the new directory
`npm install` will create /lib and pull all the required modules
`cd ..` move back out of the directory
`npm install -g meteor-up` install the app from the local directory
you should then be able to use `mup` commands as needed

ilan-schemoul commented:
This should be the top priority.

ilan-schemoul commented:
Great !

brylie commented:
Correct me if I'm wrong, but we might need to document this feature in the README (#432).

ilan-schemoul commented:
How do we enable let's encrypt support please ?
EDIT : as npm package hasn't been updated I tried to compile locally the package but I failed (see https://github.com/zodern/meteor-up/issues/98)

tim-phillips commented:
@NitroBAY let's give @zodern more than 16 hours before we complain about the npm package

zodern commented:
```
meteor: {
  ...
  ssl: {
    autogenerate: {
      email: 'email.address@domain.com',
      domains: 'website.com,www.website.com' // coma separated list of domains
    }
  }
}
```

I will update the npm package and documentation this weekend.

ilan-schemoul commented:
@babenzele I'm absolutely not complaining I wanted to notice that the compilation fails. Sorry if you think I was in any way aggressive I really think @zodern makes a great work ;)
@zodern thanks a lot, really for your reactivity and all your work

zodern commented:
The npm package is updated.
